src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -53,28 +54,54 @@ class LockGoal(Goal):
+    addresses: Addresses, pip_tools_subsystem: PipToolsSubsystem, workspace: Workspace
 ) -> LockGoal:
+    # TODO: Looking at the transitive closure to generate a single lockfile will not work when we
+    #  have multiple lockfiles supported, via per-tool lockfiles and multiple user lockfiles.
+    #  Ideally, `./pants lock ::` would mean "regenerate all unique lockfiles", whereas now it
+    #  means "generate a single lockfile based on this transitive closure."
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest(addresses))
     reqs = PexRequirements.create_from_requirement_fields(
-        tgt[PythonRequirementsField] for tgt in targets if tgt.has_field(PythonRequirementsField)
+        tgt[PythonRequirementsField]
+        # NB: By looking at the dependencies, rather than the closure, we only generate for
+        # requirements that are actually used in the project.
+        #
+        # TODO: It's not totally clear to me if that is desirable. Consider requirements like
+        #  pydevd-pycharm. Should that be in the lockfile? I think this needs to be the case when
+        #  we have multiple lockfiles, though: we shouldn't look at the universe in that case,
+        #  only the relevant subset of requirements.
+        #
+        # Note that the current generate_lockfile.sh script in our docs also mixes in
+        #  `requirements.txt`, but not inlined python_requirement_library targets if they're not
+        #  in use. We don't have a way to emulate those semantics because at this point, all we
+        #  have is `python_requirement_library` targets without knowing the source.
+        for tgt in transitive_targets.dependencies
+        if tgt.has_field(PythonRequirementsField)
     )
 
+    if not reqs:
+        logger.warning(
+            "No third-party requirements found for the transitive closure, so a lockfile will not "
+            "be generated."
+        )
+        return LockGoal(exit_code=1)
+
     input_requirements_get = Get(
         Digest, CreateDigest([FileContent("requirements.in", "\n".join(reqs).encode())])
     )
+    # TODO: Figure out named_caches for pip-tools. The best would be to share the cache between
+    #  Pex and Pip. Next best is a dedicated named_cache.
     pip_compile_get = Get(
         VenvPex,
         PexRequest(
             output_filename="pip_compile.pex",
             internal_only=True,
             requirements=PexRequirements(pip_tools_subsystem.all_requirements),
-            # TODO: Figure out which interpreter constraints to use...Among other reasons, this is
-            #  tricky because python_requirement_library targets don't have interpreter constraints!
+            # TODO: Figure out which interpreter constraints to use...Likely get it from the
+            #  transitive closure. When we're doing a single global lockfile, it's fine to do that,
+            #  but we need to figure out how this will work with multiple resolves.
```

Comment:
Discussed a bit on https://github.com/pantsbuild/pants/pull/12300#issuecomment-876796462.

For now though, this should likely use the global default interpreter_constraints to be _slightly_ more correct (although obivously not all the way there...)